test/e2e/set/scyllacluster/datainserter.go

```diff
@@ -0,0 +1,148 @@
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/gocql/gocql"
+	"github.com/scylladb/gocqlx/v2"
+	"github.com/scylladb/gocqlx/v2/table"
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const nRows = 10
+
+type DataInserter struct {
+	client            corev1client.CoreV1Interface
+	session           *gocqlx.Session
+	table             *table.Table
+	data              []*test
+	replicationFactor int
+}
+
+type test struct {
+	Id   int
+	Data string
+}
+
+func NewDataInserter(ctx context.Context, client corev1client.CoreV1Interface, sc *scyllav1.ScyllaCluster, replicationFactor ...int) (*DataInserter, error) {
+	rf := 0
+	if len(replicationFactor) > 0 {
+		rf = replicationFactor[0]
+	} else {
+		// Default to a total number of members
+		for _, r := range sc.Spec.Datacenter.Racks {
+			rf += int(r.Members)
+		}
+	}
+
+	session, err := newSession(ctx, client, sc)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &DataInserter{
+		client:  client,
+		session: session,
+		table: table.New(table.Metadata{
+			Name:    "test.test",
+			Columns: []string{"id", "data"},
+			PartKey: []string{"id"},
+		}),
+		data: func() []*test {
+			var lst []*test
+			for i := 0; i < nRows; i++ {
+				lst = append(lst, &test{Id: i, Data: rand.String(32)})
+			}
+			return lst
+		}(),
+		replicationFactor: rf,
+	}
+
+	return d, nil
+}
+
+func (d *DataInserter) Close() {
+	d.session.Close()
+}
+
+// UpdateClientEndpoints closes an existing session and opens a new one.
+// In case an error was returned, DataInserter can no Longer be used.
+func (d *DataInserter) UpdateClientEndpoints(ctx context.Context, sc *scyllav1.ScyllaCluster) error {
+	d.session.Close()
+
+	session, err := newSession(ctx, d.client, sc)
+	if err != nil {
+		return err
+	}
+	d.session = session
+
+	return nil
+}
+
+func (d *DataInserter) Write() error {
+	err := d.session.ExecStmt(`DROP KEYSPACE IF EXISTS test`)
+	if err != nil {
+		return err
+	}
+
+	err = d.session.ExecStmt(fmt.Sprintf("CREATE KEYSPACE test WITH replication = {'class' : 'NetworkTopologyStrategy','replication_factor' : %d}", d.replicationFactor))
+	if err != nil {
+		return err
+	}
+
+	err = d.session.ExecStmt(`CREATE TABLE test.test (id int primary key, data text)`)
+	if err != nil {
+		return err
+	}
+
+	for _, t := range d.data {
+		q := d.session.Query(d.table.Insert()).BindStruct(t)
+		err = q.ExecRelease()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *DataInserter) VerifyRead() error {
+	d.session.SetConsistency(gocql.All)
+
+	var res []*test
+	q := d.session.Query(d.table.SelectAll()).BindStruct(&test{})
+	err := q.SelectRelease(&res)
+	if err != nil {
+		return err
+	}
+
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Id < res[j].Id
+	})
+	if !reflect.DeepEqual(res, d.data) {
+		return fmt.Errorf("read data differs from expected data")
```

Comment:
would it be cleaner if the `DataInserter` allowed for getting expected data and read data separately? ginkgo might be able to show more diff / context it it would be the one comparing the arrays.

Reply:
changed `VerifyRead` to `Read` and added `GetExpected` in order to explicitly compare read data with expected data in `verifyScyllaCluster` using ginkgo